src/main/kotlin/com/cultureamp/eventsourcing/PGSessionAdvisoryBookmarkLock.kt

```diff
@@ -0,0 +1,53 @@
+package com.cultureamp.eventsourcing
+
+import org.jetbrains.exposed.sql.statements.api.ExposedConnection
+import org.slf4j.LoggerFactory
+import java.sql.SQLException
+
+private val logger = LoggerFactory.getLogger(PGSessionAdvisoryBookmarkLock::class.java)
+class PGSessionAdvisoryBookmarkLock(
+    private val connectionProvider: () -> ExposedConnection<Any>,
+) {
+    private var connection = connectionProvider()
+
+    fun lock(bookmarkName: String) {
+        try {
+            acquireLock(connection, bookmarkName)
+        } catch (e: Exception) {
+            logger.info("Failed to refresh session lock for $bookmarkName, resetting connection to try again", e)
+            connection = connectionProvider()
+            acquireLock(connection, bookmarkName)
+        }
+    }
+
+    private fun acquireLock(connection: ExposedConnection<Any>, bookmarkName: String) {
+        val lockTimeoutMilliseconds = 10_000
+        val key = createKey(connection, bookmarkName)
+        while (!tryLock(connection, key, lockTimeoutMilliseconds)) {
+            logger.debug("Timed out trying to acquire lock on $bookmarkName after ${lockTimeoutMilliseconds}ms. Trying again")
+        }
+    }
+
+    private fun tryLock(connection: ExposedConnection<Any>, key: Long, lockTimeoutMs: Int): Boolean {
+        try {
+            connection.prepareStatement("SET LOCAL lock_timeout = ?", false).also {
+                it[1] = "${lockTimeoutMs}ms"
+                it.executeUpdate()
+            }
+            connection.prepareStatement("SELECT pg_advisory_lock(?)", false).also {
+                it[1] = key
+                it.executeQuery()
+            }
+        } catch (e: SQLException) {
+            if (e.message.orEmpty().contains("canceling statement due to lock timeout")) {
+                return false
+            } else {
+                throw e
+            }
+        }
+        return true
+    }
+
+    // TODO this should be done gooder, perhaps by looking up the bookmark in db?
```

Comment:
Can we copy what the ruby framework does?

Reply:
We can (which would be looking up a sequence-generated column in db), however 
1) it will add to latency (because we have to hit db again) and 
2) it leaves possibility of clash with other locks since it's just a small integer
I like [Pete's idea](https://cultureamp.slack.com/archives/C058PPB8V2R/p1706041951897979) of generating a V5 uuid from bookmark name and using least/most significant Long (half the uuid) from that. WDYT?

Reply:
Yeah if it's effectively guaranteed uniqueness for a bookmark name I can't see a reason we shouldn't do that

Reply:
adding extra column + loading it in lock is going to add a fair bit more changes + complexity. What I like about this UUID approach is that we don't need any more data. But I understand your concern about diverging from event-framework